bevy_reflect: Add `Reflectable` trait

MrGVSV opened the pull request:
# Objective

When deriving `Reflect`, users will notice that their generic arguments also need to implement `Reflect`:

```rust
#[derive(Reflect)]
struct Foo<T: Reflect> {
  value: T
}
```

This works well for now. However, as we want to do more with `Reflect`, these bounds might need to change. For example, to get #4154 working, we likely need to enforce the `GetTypeRegistration` trait. So now we have:

```rust
#[derive(Reflect)]
struct Foo<T: Reflect + GetTypeRegistration> {
  value: T
}
```

Not great, but not horrible. However, we might then want to do something as suggested in [this](https://github.com/bevyengine/bevy/issues/5745#issuecomment-1221389131) comment and add a `ReflectTypeName` trait for stable type name support. Well now we have:

```rust
#[derive(Reflect)]
struct Foo<T: Reflect + GetTypeRegistration + ReflectTypeName> {
  value: T
}
```

Now imagine that for even two or three generic types. Yikes!

As the API changes it would be nice if users didn't need to manually migrate their generic type bounds like this.

A lot of these traits are (or will/might be) core to the entire reflection API. And although `Reflect` can't add them as supertraits for object-safety reasons, they are still indirectly required for things to function properly (manual implementors will know how easy it is to forget to implement `GetTypeRegistration`). And they should all be automatically implemented for user types anyways as long they're using `#[derive(Reflect)]`.

## Solution

Add a "catch-all" trait called `Reflectable` whose supertraits are a select handful of core reflection traits.

This allows us to consolidate all the examples above into this:

```rust
#[derive(Reflect)]
struct Foo<T: Reflectable> {
  value: T
}
```

And as we experiment with the API, users can rest easy knowing they don't need to migrate dozens upon dozens of types. It should all be automatic!

## Discussion

1. Thoughts on the name `Reflectable`? Is it too easily confused with `Reflect`? Or does it at least accurately describe that this contains the core traits? If not, maybe `BaseReflect`?

---

## Changelog

- Added the `Reflectable` trait

alice-i-cecile commented on examples/reflection/generic_reflection.rs:
```diff
@@ -12,8 +12,15 @@ fn main() {
+/// Until the reflection API stabilizes, these trait bounds are liable to change. While we currently
+/// only require [`Reflect`] right now, we may eventually require others.
+///
+/// So rather than adding each trait to our generic arguments manually (and having to update them
```
```suggestion
/// Rather than adding each trait to our generic arguments manually (and having to update them
```

alice-i-cecile commented on crates/bevy_reflect/src/lib.rs:
```diff
@@ -55,6 +55,34 @@ pub use type_info::*;
 
+/// A catch-all trait bound by the core reflection traits for easy reflection-based trait bounds.
+///
+/// > __Note:__ You do _not_ need to implement this trait manually.
```
```suggestion
/// You do _not_ need to implement this trait manually.
```

No need to go crazy on the emphasis :)

MrGVSV replied:
Yeah thanks for reeling me back haha I tend to go overboard 😅

alice-i-cecile commented:
I'm in favor of this design: I've seen this sort of supertrait be super useful with `Num`.

Name is fine. Only blocking request is that this should be linked to from the `Reflect` docs (and any other traits that this subsumes).

MrGVSV commented on crates/bevy_reflect/src/type_info.rs:
```diff
@@ -8,6 +8,8 @@ use std::any::{Any, TypeId};
 /// This trait is automatically implemented by the `#[derive(Reflect)]` macro
 /// and allows type information to be processed without an instance of that type.
 ///
+/// As a core trait to the reflection system, it is a supertrait of [`Reflectable`].
```
@alice-i-cecile I'm not really sure I love this. Any thoughts or suggestions?

afonsolage reviewed (commented):
Does advising users to make use of `Reflectable` will likely increas the number of breaking changes, whenever we update `Reflectable`? I mean, if I add a new trait to `Reflectable`, and it isn't an auto impl trait, this means the user will have to add a new `#[derive]` for all types that uses the generic bounds, right?

Also, should we consider using `Reflectable` as temporary until [`trait_alias`](https://github.com/rust-lang/rust/issues/41517) stabilize (which may take years)?

MrGVSV commented:
> Does advising users to make use of `Reflectable` will likely increas the number of breaking changes, whenever we update `Reflectable`? I mean, if I add a new trait to `Reflectable`, and it isn't an auto impl trait, this means the user will have to add a new `#[derive]` for all types that uses the generic bounds, right?

`Reflect` has the restriction that it needs to be object-safe. Because of this, we have a lot of other required (or partially-required) traits to provide other functionality. `Typed` is definitely required due to `Reflect::get_type_info`, while `GetTypeRegistration` is required if you need to use the registry. However, not implementing `GetTypeRegistration` can lead to subtle footguns.

`Reflectable` is meant to unify these _core_ traits. And because these traits are so heavily utilized within bevy_reflect, they are automatically derived via `#[derive(Reflect)]`.

I don't think `Reflectable` should contain anything that isn't auto-implemented via that derive macro. Therefore, adding a trait should only be a breaking change for those subset of users that implement `Reflect` (et al.) manually— which probably isn't many haha.

What _could_ be a breaking change, though, is _removing_ a trait. Anyone that used `Reflectable` to ensure a trait impl existed would suddenly not have that guarantee anymore. But if we removed a trait from `Reflectable`, we likely also removed it from the derive macro, meaning that this would still be a breaking change with-or-without `Reflectable`.


> Also, should we consider using `Reflectable` as temporary until [`trait_alias`](https://github.com/rust-lang/rust/issues/41517) stabilize (which may take years)?

I don't know much about it, but it looks like pretty much what we're trying to do. However, I'm not too sure what advantages it has over the blanket impl approach.

alice-i-cecile commented:
> Reflectable is meant to unify these core traits. And because these traits are so heavily utilized within bevy_reflect, they are automatically derived via #[derive(Reflect)].

We should consider making the derive name the same as the name of the blanket trait. IMO that means:

1. `Reflect` trait name is used for this `Reflectable` trait.
2. The current `Reflect` functionality becomes something like `BasicReflect`.

MrGVSV commented:
> We should consider making the derive name the same as the name of the blanket trait. IMO that means:
> * `Reflect` trait name is used for this `Reflectable` trait.
> * The current `Reflect` functionality becomes something like `BasicReflect`.

I don't mind this as it makes it slightly clearer that it will be deriving multiple traits (at least, if a user knows what `Reflectable` is). Or, I at any rate I suppose it brings the trait to the attention of users.

However, I don't really like the idea of renaming `Reflect` to `BaseReflect`. All the other supertraits of `Reflectable` help drive the other parts of reflection. But most of that is utilized _within_ bevy_reflect. Apart from generic types and functions where those bounds are necessary, users generally won't need to worry about them.

Most often, they'll just be working with `dyn Reflect` as opposed to a concrete generic type. As it stands right now, this PR shouldn't break anything on its own. If we changed to `BaseReflect` then the only thing users will see from it is that they need to use `dyn BaseReflect` now.

If we wanted to change anything, I feel it should at most be in your suggestion to make this the trait we "derive" via `#[derive(Reflectable)]`. But I think that could be saved for a more controversial followup PR haha.

brandon-reinhart reviewed (commented):
This looks straightforward to me.

Is FromReflect also needed?

At the moment, I need:

```rust
pub trait EntityLocator:
    GetTypeRegistration + Typed + FromReflect + Reflect + TypePath
```

To do this:

```rust
    fn register_entity_locator<Locator: EntityLocator>(&mut self) -> &mut Self {
        self.register_type::<MapLocations<Locator>>();
        
        ... other stuff
      }
```

Without which I get:

```rust
error[E0277]: `Locator` can not be created through reflection
   --> game\tier1\src\ai\locations\map_locations.rs:11:30
    |
11  |         self.register_type::<MapLocations<Locator>>();
    |              -------------   ^^^^^^^^^^^^^^^^^^^^^ the trait `bevy::prelude::FromReflect` is not implemented for `Locator`, which is required by `MapLocations<Locator>: bevy::bevy_reflect::GetTypeRegistration`
    |              |
    |              required by a bound introduced by this call
    |
    = note: consider annotating `Locator` with `#[derive(FromReflect)]`
```

MrGVSV commented:
> This looks straightforward to me.
> 
> Is FromReflect also needed?

Good question! I chose to leave `FromReflect` out since it's not strictly required by reflection. It can be opted out of, which would make `Reflectable` a lot more restrictive than would probably be ideal.

So in the cases you need `FromReflect`, I'd suggest just making that part explicit:

```rust
pub trait EntityLocator: Reflectable + FromReflect {}
```

alice-i-cecile commented on crates/bevy_reflect/src/reflectable.rs:
```diff
@@ -0,0 +1,33 @@
+use crate::{GetTypeRegistration, Reflect, TypePath, Typed};
+
+/// A catch-all trait that is bound by the core reflection traits,
+/// useful for simpler reflection-based generic type bounds.
```
```suggestion
/// useful to simplify reflection-based generic type bounds.
```

alice-i-cecile reviewed (approved):
I have a wording suggestion, but I think this is useful and clear.

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1687 if you'd like to help out.
